db/collection.go

```diff
@@ -196,13 +211,108 @@ func (c *Collection) Find(q *Query, opts ...TxnOption) (instances [][]byte, err
 	return
 }
 
+// validInstance validates the json object against the collection schema.
+func (c *Collection) validInstance(v []byte) error {
+	r, err := gojsonschema.Validate(c.schemaLoader, gojsonschema.NewBytesLoader(v))
+	if err != nil {
+		return err
+	}
+	errs := r.Errors()
+	if len(errs) == 0 {
+		return nil
+	}
+	var msg string
+	for i, e := range errs {
+		msg += e.Field() + ": " + e.Description()
+		if i != len(errs)-1 {
+			msg += "; "
+		}
+	}
+	return fmt.Errorf("%w: %s", ErrInvalidSchemaInstance, msg)
+}
+
+// validWrite validates new events against the identity and user-defined write validator function.
+func (c *Collection) validWrite(identity thread.PubKey, e core.Event) error {
+	if c.writeValidator == nil {
+		return nil
+	}
+	vm := goja.New()
+	validate, writer, err := getJSFunc(vm, c.writeValidator, identity)
+	if err != nil {
+		return err
+	}
+	data, err := e.Marshal()
+	if err != nil {
+		return err
+	}
+	event, err := parseJSON(vm, data)
+	if err != nil {
+		return err
+	}
+	var inv goja.Value
+	val, err := c.db.datastore.Get(baseKey.ChildString(c.name).ChildString(e.InstanceID().String()))
+	if err != nil && !errors.Is(err, ds.ErrNotFound) {
+		return err
+	}
+	if val != nil {
+		inv, err = parseJSON(vm, val)
+		if err != nil {
+			return err
+		}
+	}
+	res, err := validate(nil, writer, event, inv)
+	if err != nil {
+		return fmt.Errorf("error running write validator func: %v", err)
+	}
+	out := res.Export()
+	switch out.(type) {
+	case bool:
+		if out.(bool) {
+			return nil
+		} else {
+			return app.ErrInvalidNetRecordBody
+		}
+	case nil:
+		return app.ErrInvalidNetRecordBody
+	default:
+		return fmt.Errorf("%w: %v", app.ErrInvalidNetRecordBody, out)
```

Comment:
Maybe could be merged with the `nil` case? 
(Maybe not if `nil` case describes some other case I'm missing)

Reply:
I kept them separate to avoid having to do some other checking to construct the error message. If the function returned an object of any kind, a `number`, or a `string`, that will be used as the error message. If they returned `false` or `undefined`, they just get the default error message (no wrapping).

Reply:
Cool, same "undefined" thing as above.